pkg/network/tracer/connection/ebpfless/tcp_processor.go

```diff
@@ -0,0 +1,338 @@
+	hasSentPacket bool
+	// localStartSeq is the initial outgoing tcp.Seq - only used for debugging relative seqs
+	localStartSeq uint32
+	// maxSeqSent is the latest outgoing tcp.Seq if hasSentPacket==true
+	maxSeqSent uint32
+	// hasAckedPacket is whether the outgoing side has ACK'd any packets
+	hasAckedPacket bool
+	// lastAck is the latest outgoing tcp.Ack if hasAcked==true
+	lastAck uint32
+
+	// hasReceivedPacket is whether anything has been received - only used for debugging (to see if remoteStartSeq exists)
+	hasReceivedPacket bool
+	// remoteStartSeq is the initial incoming tcp.Seq - only used for debugging
+	remoteStartSeq uint32
+
+	// hasSynAcked is used to keep track of whether we should advance to TcpStateEstablished
+	// (because the state transition requires multiple packets)
+	hasSynAcked bool
+
+	// hasLocalFin is whether the outgoing side has FIN'd
+	hasLocalFin bool
+	// hasRemoteFin is whether the incoming side has FIN'd
+	hasRemoteFin bool
+	// localFinSeq is the tcp.Seq number for the outgoing FIN (including any payload length)
+	localFinSeq uint32
+	// remoteFinSeq is the tcp.Seq number for the incoming FIN (including any payload length)
+	remoteFinSeq uint32
+}
+
+type TCPProcessor struct {
+	conns map[network.ConnectionTuple]connectionState
+}
+
+func NewTCPProcessor() *TCPProcessor {
+	return &TCPProcessor{
+		conns: map[network.ConnectionTuple]connectionState{},
+	}
+}
+
+// Process handles a TCP packet, calculating stats and keeping track of its state according to the
+// TCP state machine.
+// https://users.cs.northwestern.edu/~agupta/cs340/project2/TCPIP_State_Transition_Diagram.pdf
+func (t *TCPProcessor) Process(conn *network.ConnectionStats, pktType uint8, ip4 *layers.IPv4, ip6 *layers.IPv6, tcp *layers.TCP) error {
+	payloadLen, err := TCPPayloadLen(conn.Family, ip4, ip6, tcp)
+	if err != nil {
+		return err
+	}
+
+	st := t.conns[conn.ConnectionTuple]
+	log.TraceFunc(func() string {
+		return fmt.Sprintf("pre ack_seq=%+v", st)
+	})
+
+	payloadSeq := tcp.Seq + uint32(payloadLen)
+
+	switch pktType {
+	case unix.PACKET_OUTGOING:
+		conn.Monotonic.SentPackets++
+		if !st.hasSentPacket {
+			st.localStartSeq = tcp.Seq
+		}
+		if !st.hasSentPacket || isSeqBefore(st.maxSeqSent, payloadSeq) {
+			st.hasSentPacket = true
+			conn.Monotonic.SentBytes += uint64(payloadLen)
+			st.maxSeqSent = payloadSeq
+		} else {
+			// TODO retransmit
+		}
+		if tcp.ACK {
+			isFinAck := st.hasRemoteFin && tcp.Ack == st.remoteFinSeq+1
+			if !st.hasAckedPacket {
+				st.hasAckedPacket = true
+				st.lastAck = tcp.Ack
+			} else if isSeqBefore(st.lastAck, tcp.Ack) {
+				ackDiff := tcp.Ack - st.lastAck
+				// if this is ack'ing a fin packet, there is an extra sequence number to cancel out
+				if isFinAck {
+					ackDiff--
+				}
+				conn.Monotonic.RecvBytes += uint64(ackDiff)
+				st.lastAck = tcp.Ack
+			}
+		}
+	case unix.PACKET_HOST:
+		conn.Monotonic.RecvPackets++
+		if !st.hasReceivedPacket {
```

Comment:
You could also just use `conn.Monotonic.RecvPackets` here.

Reply:
I added `hasReceivedPacket` under the impression that `RecvPackets` would remain nonzero/keep climbing if a connection tuple is reopened a second/third time in between network checks. Network checks drain out the `ConnectionStats` structure right?